docs/source/rabbitmq-tutorial/examples/7-publisher-confirms/publish_asynchronously.py

```diff
@@ -0,0 +1,54 @@
+import asyncio
+
+from aio_pika import Message, connect
+
+from aiormq.exceptions import DeliveryError
+
+
+def get_messages_to_publish():
+    for i in range(10000):
+        yield f"Hello World {i}!".encode()
+
+
+def handle_confirm(confirmation):
+    try:
+        _ = confirmation.result()
+        # code when message is ack-ed
+    except DeliveryError:
+        # code when message is nack-ed
+        pass
+    except TimeoutError:
+        # code for message timeout
+        pass
+    else:
+        # code when message is confirmed
+        pass
+
+
+async def main() -> None:
+    # Perform connection
+    connection = await connect("amqp://guest:guest@localhost/")
+
+    async with connection:
+        # Creating a channel
+        channel = await connection.channel()
+
+        # Declaring queue
+        queue = await channel.declare_queue("hello")
+
+        async with asyncio.TaskGroup() as tg:
+            # Sending the messages
+            for msg in get_messages_to_publish():
+                tg.create_task(
+                    channel.default_exchange.publish(
+                        Message(msg),
+                        routing_key=queue.name,
+                        timeout=5.0,
+                    )
+                ).add_done_callback(handle_confirm)
+
+        print(" [x] Sent and confirmed multiple messages asynchronously. ")
```

Comment:
```suggestion
        # List for async tasks
        tasks = []

        # Sending the messages
        for msg in get_messages_to_publish():
            task = asyncio.create_task(
                channel.default_exchange.publish(
                    Message(msg),
                    routing_key=queue.name,
                    timeout=5,
                )
            )
            tasks.append(task)

        results = await asyncio.gather(*tasks, return_exceptions=True)

        success = 0
        error = 0
        for result in results:
            if isinstance(result, (DeliveryError, asyncio.TimeoutError)):
                error += 1
            else:
                success += 1

        print(" [x] Sent and confirmed multiple messages asynchronously. ")
        print("Success:", success)
        print("Errored:", error)
```

Reply:
I don't think this solves the problem. This is more like a variation of the batch confirmation, except it sends all messages as a single batch. It also has no way to determine which messages failed or succeeded. It only ever detects failures after all messages have been published, so it cannot react to delivery errors or confirmations asynchronously. It only works for a fixed amount of messages and not for sending an infinite amount of messages.

So I don't think this code suggestion matches the ideas that are presented as the asynchronous "Strategy 3" in the original [RabbitMQ tutorial](https://www.rabbitmq.com/tutorials/tutorial-seven-java.html)

I suppose with the new `get_underlay_channel()` method one could use `channel.get_underlay_channel().delivery_tag` to get the latest delivery tag:
```python

awaiting_confirmations = {}
delivery_tag_var = ContextVar('delivery_tag')
#...

for msg in get_messages_to_publish():
    task = asyncio.create_task(
        channel.default_exchange.publish(
            Message(msg),
            routing_key=queue.name,
            timeout=5,
        )
    )
    delivery_tag = channel.get_underlay_channel().delivery_tag #Get last used delivery tag from underlying channel
    delivery_tag_var.set(delivery_tag) #Required, so we can also know the delivery_tag in case of Timeout. For DeliveryError we can get it from the frame contained in the exception
    awaiting_confirmations[delivery_tag] = msg #Store message body to maybe resend
    task.add_done_callback(handle_confirms)
```

What do you think?

Reply:
`asyncio.create_task` plans a task and starts executing it at the first context switch, just like in javascript, if you write `while () {}` it won't continue because the context switch won't happen.

Even if you try to make `StreamWriter.write` be called as soon as possible, it is just writing data to the buffer, it may not be sent over the network if the event loop is blocked (the same `while 1: pass`), this really like in the JS, but in that case the event-loop in implicitly hideen by JS-runtime.

That's the way the world works, not perfectly, of course.